server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    # now check each "good" controller and get the tarballs it contains
+    for controller in hier.controllers:
+        direct_entries = glob.iglob(os.path.join(hier.path, controller, "*"))
+        hidden_entries = Path(hier.path, controller).glob(".*")
+        if hidden_entries:
+            for hid_entries_path in hidden_entries:
+                if hid_entries_path.is_file():
+                    hier.add_unexpected_entries(
+                        Hierarchy.UNEXPECTED_OBJECTS, controller, hid_entries_path.name,
+                    )
+        controller_subdir = list()
+        for item in direct_entries:
+            item_path = Path(item)
+            if item_path.is_dir():
+                controller_subdir.append(item_path.name)
+            elif item_path.is_symlink():
+                try:
+                    item_p = item_path.resolve(strict=True)
+                except FileNotFoundError:
+                    hier.add_error_or_inconceivable_entries(
+                        Hierarchy.ERROR,
+                        controller,
+                        f"{item_path.name}_path, '{item_path}', does not resolve to a real location",
+                    )
+                symlink_item = f"{item_path.name} -> {item_p}"
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_SYMLINKS, controller, symlink_item
+                )
+            elif all(
+                [
+                    item_path.is_file(),
+                    not item_path.name.endswith(".tar.xz"),
+                    not item_path.name.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_OBJECTS, controller, item_path.name
+                )
+            elif item_path.is_file() and (
+                item_path.name.endswith(".tar.xz")
+                or item_path.name.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                hier.add_error_or_inconceivable_entries(
+                    Hierarchy.INCONCEIVABLE, controller, item_path.name,
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return 0
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
+    for tb in tarball_dirs:
+        if tb.endswith("unpack"):
+            tar = tb[:-7]
+            tar = f"{tar}.tar.xz"
+            dict_val = Hierarchy.INVALID_UNPACKING_DIRS
+        else:
+            tar = f"{tb}.tar.xz"
+            dict_val = Hierarchy.INVALID_TB_DIRS
+        tarfile = Path(ihier.config.ARCHIVE, controller, tar)
+        if tarfile.exists():
+            if os.access(tarfile, os.R_OK):
+                continue
+        else:
+            tblist(dict_val, controller, tb)
```

Comment:
I'm a little confused by this code:
 - if the tarball doesn't exist, then we hand it to the callback
 - if exists and is accessible, then we move on to the next file
 - but, if it exists and is inaccessible, then we drop it into a crack?

I _think_ the following code would be better:
```
        if not os.access(tarfile, os.R_OK):
            ihier.add_unexpected_entries(dict_val, controller, tb)
```
If the tar file is unreadable (e.g., because it doesn't exist or because it's protected), then record it as an unexpected entry; otherwise, continue to the bottom of the loop and on to the next entry.